connectors/byoc.py

```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
-        self.created_at = datetime.now(timezone.utc)
-        self.completed_at = None
-        self.filtering = Filter()
-        self.client = elastic_index.client
+class ESDocument:
+    def __init__(self, elastic_index, doc_source=None):
+        self.index = elastic_index
         if doc_source is None:
             doc_source = dict()
+        self.id = doc_source.get("_id")
+        self._source = doc_source.get("_source", {})
 
-        self.doc_source = doc_source
-        self.job_id = self.doc_source.get("_id")
-        self.status = str2e(self.doc_source.get("_source", {}).get("status"), JobStatus)
+    def get(self, *keys, default=None):
+        value = self._source
+        for key in keys:
+            if not isinstance(value, dict):
+                return default
+            value = value.get(key)
+        if value is None:
+            return default
+        return value
 
+    async def reload(self):
+        return await self.index.fetch_by_id(self.id)
+
+
+class SyncJob(ESDocument):
     @property
-    def duration(self):
-        if self.completed_at is None:
-            return -1
-        msec = (self.completed_at - self.created_at).microseconds
-        return round(msec / 9, 2)
+    def status(self):
+        return str2e(self.get("status"), JobStatus)
+
+    @property
+    def error(self):
+        return self.get("error")
+
+    @property
+    def connector_id(self):
+        return self.get("connector", "id")
 
     @property
     def index_name(self):
-        return self.doc_source["_source"]["connector"]["index_name"]
+        return self.get("connector", "index_name")
 
-    async def start(self, trigger_method=JobTriggerMethod.SCHEDULED, filtering=None):
-        if filtering is None:
-            filtering = Filter()
+    @property
+    def language(self):
+        return self.get("connector", "language")
 
-        self.status = JobStatus.IN_PROGRESS
-        self.filtering = filtering
+    @property
+    def service_type(self):
+        return self.get("connector", "service_type")
 
-        job_def = {
-            "connector": {
-                "id": self.connector_id,
-                "filtering": SyncJob.transform_filtering(filtering),
-            },
-            "trigger_method": e2str(trigger_method),
-            "status": e2str(self.status),
-            "error": None,
-            "deleted_document_count": 0,
-            "indexed_document_count": 0,
-            "created_at": iso_utc(self.created_at),
-            "completed_at": None,
-        }
-        resp = await self.client.index(index=JOBS_INDEX, document=job_def)
-        self.job_id = resp["_id"]
-        return self.job_id
+    @property
+    def configuration(self):
+        return DataSourceConfiguration(self.get("connector", "configuration"))
 
-    async def done(self, indexed_count=0, deleted_count=0, exception=None):
-        self.completed_at = datetime.now(timezone.utc)
+    @property
+    def filtering(self):
```

Comment:
Shouldn't `filtering` be of type array in a sync job? (see: [Protocol](https://github.com/elastic/connectors-ruby/blob/main/docs/CONNECTOR_PROTOCOL.md)) Btw I think we need to update these docs and maybe add them also to connectors-python.

Reply:
Yes, that would be a problem. Also the new sync rule object model will probably use multiple filters for different tables. Therefore it definitely should be an array.